client/state/imports/reducer.js

```diff
@@ -1,13 +1,182 @@
+
 /**
  * Internal dependencies
  */
-import { combineReducers, withStorageKey } from 'calypso/state/utils';
-import uploadsReducer from 'calypso/state/imports/uploads/reducer';
-import siteImporterReducer from 'calypso/state/imports/site-importer/reducer';
+import {
+	IMPORTS_AUTHORS_SET_MAPPING,
+	IMPORTS_AUTHORS_START_MAPPING,
+	IMPORTS_IMPORT_CANCEL,
+	IMPORTS_IMPORT_LOCK,
+	IMPORTS_IMPORT_RECEIVE,
+	IMPORTS_IMPORT_RESET,
+	IMPORTS_IMPORT_START,
+	IMPORTS_IMPORT_UNLOCK,
+	IMPORTS_START_IMPORTING,
+	IMPORTS_UPLOAD_FAILED,
+	IMPORTS_UPLOAD_COMPLETED,
+	IMPORTS_UPLOAD_SET_PROGRESS,
+	IMPORTS_UPLOAD_START,
+} from 'calypso/state/action-types';
+import { combineReducers, keyedReducer, withStorageKey } from 'calypso/state/utils';
+import uploads from './uploads/reducer';
+import siteImporter from './site-importer/reducer';
+import { appStates } from './constants';
+import { fromApi } from './api';
+
+function isImporterStatusHydrated( state = false, action ) {
+	switch ( action.type ) {
+		case IMPORTS_IMPORT_RECEIVE:
+			return true;
+	}
+
+	return state;
+}
+
+function importerStatus( state = {}, action ) {
+	switch ( action.type ) {
+		case IMPORTS_IMPORT_CANCEL:
+		case IMPORTS_IMPORT_RESET:
+			return omit( state, action.importerId );
+
+		case IMPORTS_UPLOAD_FAILED:
+			return {
+				...state,
+				[ action.importerId ]: {
+					...state[ action.importerId ],
+					importerState: appStates.UPLOAD_FAILURE,
+					errorData: { type: 'uploadError', description: action.error },
+				},
+			};
+
+		case IMPORTS_UPLOAD_COMPLETED:
+			return {
+				...omit( state, action.importerId ),
+				[ action.importerStatus.importerId ]: action.importerStatus,
+			};
+
+		case IMPORTS_IMPORT_RECEIVE: {
+			// The REST endpoint returns an empty `[]` array if there are no known imports.
+			// In that case we don't update the `importerStatus` map, and only mark it as hydrated.
+			if ( isEmpty( action.importerStatus ) ) {
+				return state;
+			}
+
+			// don't receive the response if the importer is locked
+			if ( action.isLocked ) {
+				return state;
+			}
+
+			// convert the response with `fromApi` only after we know it's not empty
+			const newImporterStatus = fromApi( action.importerStatus );
+
+			return omitBy(
+				{
+					// filter the original set of importers...
+					...state,
+					// ...and the importer being received.
+					[ newImporterStatus.importerId ]: newImporterStatus,
+				},
+				( importer ) =>
+					[ appStates.CANCEL_PENDING, appStates.DEFUNCT, appStates.EXPIRED ].includes(
+						importer.importerState
+					)
+			);
+		}
+
+		case IMPORTS_AUTHORS_START_MAPPING:
+			return {
+				...state,
+				[ action.importerId ]: {
+					...state[ action.importerId ],
+					importerState: appStates.MAP_AUTHORS,
+				},
+			};
+
+		case IMPORTS_AUTHORS_SET_MAPPING:
+			return {
+				...state,
+				[ action.importerId ]: {
+					...state[ action.importerId ],
+					customData: {
+						...state[ action.importerId ].customData,
```

Comment:
Is it possible `.customData` won't exist?

Reply:
If the `IMPORTS_AUTHORS_SET_MAPPING` is called at the right time, i.e., after the import file has been uploaded and analyzed, then yes, `customData` should always be there.

Even if it wasn't present, the reducer doesn't crash and returns a well-formed state object. Because spreading an undefined value in object is a valid operation (unlike in array where it's error).

The reducer could be tidied up a little more when removing Lodash's `get` and `map` usages.